site-src/geps/gep-1282.md

```diff
@@ -98,12 +106,297 @@ Of course, from the section above, you can see that there are many other feature
 
 ## API
```

Comment:
I mentioned this on the call but want to re-iterate so its in writing: I think BackendCapabilities is the wrong abstraction for this API.

What `BackendCapabilities` says is "Here are the properties of the backend". This may be something like "Listens on TLS1.2 or TLS1.3 over HTTP/2" or "Supports HTTP/1.1 and Http/2", etc.

What we want to do, as implementors, is _program a client_ to decide how to connect. With the information in this API, we don't have enough data to know how to program a client.

If a backend says they support HTTP/1.1 and HTTP/2, how do I know which one to send? 

If a backend says it supports TLS, I have the same problem. As a gateway, I can either forward a request as is or add a layer of TLS. But in many cases, we won't actually know if the original request already *is* TLS encrypted, so we don't know whether we should add a layer of TLS or not.

Ultimately, this configuration is a *consumer configuration*; by treating it as a producer configuration we are not getting the functionality required to implement this effectively.

Note: you may look at DestinationRule and make the same argument, but actually it is a consumer configuration that allows the producer to set defaults for consumers; for reasons mentioned above, setting a default policy of "TLS encrypt" would almost certainly break users.

Reply:
Thanks @howardjohn for writing this out, I understand the concern better.  

My viewpoint, and I think @youngnick 's, is that by creating `BackendProperties` that reference a backend, the developer or admin is telling the Gateway how to interact with the backend.  The intention is still to use `BackendCapabilities` as a wrapper or decorator for the backend.  

I may miss some nuances here, but I thought what might happen is that the `Gateway` (acting in its client role) would lookup and use `BackendCapabilites` as its reference to the backend. For a `HTTPRoute` that is not a passthrough (assuming the `Gateway` knows which traffic is passthrough), if the `Gateway` finds a `BackendCapabilitesSpec` for the backend, then it uses that information to decide how to  connect.  If it doesn't find a spec, then it connects as it would have without knowing the backend capabilities.

Note I started with `BackendTargetRef` as a part of `BackendCapabilities`, not `BackendCapabilitesSpec`.  Later I moved it into the spec by request from @robscott (https://github.com/kubernetes-sigs/gateway-api/pull/1430#discussion_r987472884).  Does hiding the `BackendTargetRef` within `BackendCapabilitesSpec` make programming a client more difficult?  

Is there any way to salvage the `BackendCapabilies` as an abstraction?  We can definitely add more to make it work properly if it is still missing too much.  Maybe we need to store more of the handshake information for TLS, there are a few pieces missing, like details on cipher suite, compression, curve, etc.  And there are few pieces that still need tweaking.

Reply:
@howardjohn, there are a few reasons why we're doing this with a separate resource:
- To avoid adding fields to the `backendRef` that may not always make sense there
- To avoid having to set fields individually for each `backendRef` - this cost increases for each time a backend is referenced in a Route config.
- To make it so that this config is available to other things than HTTPRoute without reimplementing in every Route type

The basic idea is that there's an object, owned by the _owner of the Service_ that tells consumers what they can do when they connect to the service. You're one hundred percent right that these are _consumer_ rules, but the only person who knows what these rules should be is the person who sets up the Service - the Producer.. So this is a way for the _service owner/producer_ to tell _Service consumers_ about how to connect to the service, in a structured way.

It seems to me like having the Route/Gateway need to guess at these rules is not ideal, unless we assume that the Route owner _always_ is the Service producer. 

If that is the case, then we can easily do this by adding fields to `backendRef`. The tradeoff there is that now, _everywhere_ that uses `backendRef` needs to have rules about what happens when you specify TLS settings, or websocket settings, or whatever. So if you use your backend in more than one backendRef, you now need to duplicate all these settings.

Reply:
Is the problem just that calling this "backend properties/capabilities" is guiding towards wrong abstraction? For example, the description currently in the TLS fields says:

```go
    // The minimal TLSVersion that is accepted by the backend.
    ...
    // CACerts are the Certificate Authority, trusted by both Gateway and backend,
```

This wording makes sense if one would really describe _capabilities of the backend_. But since we are programming the client/gateway/proxy, speculating about the backend capabilities like this seems bit unexpected. 

What I was expecting to see was simply properties applied to Gateway's _backend connection(s)_ when communicating towards specific backend(s). Something like "the minimum TLSVersion used by the Gateway when connecting to the backend" and "Certificate Authorities used by Gateway to verify backend server certificate".  I'm not sure if the same holds to everything, but at least for TLS it seems so.

Reply:
The alternative is not to make _every use in `backendRef`_. The core relationship of consumer + producer rules, IMO, is "Producer sets defaults, can be overriden by consumer" (which is essentially what policy attachment is all about, as well). Its not necessarily either-or.

---

> For a HTTPRoute that is not a passthrough (assuming the Gateway knows which traffic is passthrough), if the Gateway finds a BackendCapabilitesSpec for the backend, then it uses that information to decide how to connect.

Focusing overly on HTTPRoute (here and in other GEPs, we tend to) is not always the best path. What if the route is a TCPRoute? The contents are opaque - we cannot tell if its TLS or not.

Even if its a TLSRoute - you know it has _a layer of TLS_, but how many? How many does the backend expect? May sound crazy, but its not uncommon to have multiple layers of TLS at this point with transparent mTLS, things like [onion routing](https://en.wikipedia.org/wiki/Onion_routing), etc.

But most importantly - backendProperties, as proposed, are a property of the backend. That means there is no relationship to Gateway or Routes _at all_. These properties should be respected by any Kubernetes cluster that can support them, not just N/S traffic (mesh).

Reply:
> What I was expecting to see was simply properties applied to Gateway's backend connection(s) when communicating towards specific backend(s). Something like "the minimum TLSVersion used by the Gateway when connecting to the backend" and "Certificate Authorities used by Gateway to verify backend server certificate". I'm not sure if the same holds to everything, but at least for backend TLS hop it seems so.

Is that not what this proposal is doing? Is it just the _name_ `BackendCapabilities` that is the sticking point? The idea here is to allow the _owner of the service_ to say "When connecting to the _thing that's running inside the pod_. do these things" - have a minimum TLS version, use this CA cert and so on. If it's just that the word `Capabilities` is too overloaded, and we need to change the name, I don't think that's a big deal.

@tsaarni, in the quote you've given, I think it's equally correct to have that say:
```
    // The minimal TLSVersion required to connect to the backend.
    ...
    // CACerts are the Certificate Authority, that the serving certificate for the backend service must be validated against.
```

Both of those properties are things that are _set_ by the owner of the Service, but are _needed_ by clients that connect to the Service. Similarly, the point of having websocket details in a resource like this is so that the owner of the Service can say "My service will accept websockets at the following paths", so that clients (like a Gateway proxy) know what to do for those paths.

Again, the point here is to take a whole bunch of configuration that, in the Ingress world, was often done via annotations on the Service, and make it structured. Ideally, we would put this inside the Service in structured fields, but we can't easily change the Service spec, so we've opted for a "decorator" style CRD.

If it's just that the _name_ `BackendCapabilities` is creating the wrong abstraction, that's fine, but we need some guidance around what else to call it.

@howardjohn, I'm not quite sure I understand what you're asking for. Are you saying that this resource should be called BackendHTTPProperties or something, and only be relevant when it's used as a `backendRef` in a HTTPRoute? Or are you saying that we should define semantics for the other Route types?

It's clear that you disagree with the current approach, but I am not sure what actions we need to take to move towards something we can agree on.

Reply:
Consumers need to be able to control the settings, not just the service producers.

I am definitely not suggesting making it tied to HTTP - my concern was the opposite. That this won't work for non-http clients because we don't know if we should encrypt or not.

Reply:
Okay, I agree that consumers also need a place for some settings like this. But don't we need to have the service producer's settings available at least, and then we can see what things make sense to allow in other places? I'll be honest, in my head I kind of thought that Policy Attachment could be used to set things from the Gateway side - so the producer can say "I accept TLS 1.2 or greater", and a Policy could say "All Routes attached to this Gateway will default to TLS 1.3" and then things would work. But without the bit where the producer says what they expect, then you can't even do simple cases, where there's no extra TLS wrapping etc.

Reply:
I agree with you but I don't think the GEP behaves that way? Or I read it wrong

Reply:
I don't think we have anything in the GEP right now about Policy Attachment and consumer settings, but that's probably worth adding.

I'll review the GEP again and see if I can see any other changes, and then come back here.